### PR TITLE
Pushdown CAST in JOIN criteria for PostgreSql

### PR DESCRIPTION
**Pull request by ssheikin**
# To be revisited after https://github.com/trinodb/trino/pull/22728 is merged

## Description

Based and requires https://github.com/trinodb/trino/pull/21607
Similar to https://github.com/trinodb/trino/pull/22728

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java [hunk 1]
```diff
@@ -1013,6 +1012,66 @@ public void testInPredicatePushdown()
         }
     }
 
+    @Test
```
Comment (ssheikin) [inline plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:1015]: Add case sensitive tests

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java [hunk 2]
```diff
@@ -1013,6 +1017,68 @@ public void testInPredicatePushdown()
...
+        CastTestCase testCaseMoney = new CastTestCase("money", "varchar(100)", "money");
+        try (TestTable tableA = new TestTable(onRemoteDatabase(), "test_join_cast_a_", "(a1 %s, a2 %s)".formatted(testCaseMoney.fromRemoteType(), testCaseMoney.fromRemoteType()));
+                TestTable tableB = new TestTable(onRemoteDatabase(), "test_join_cast_b_", "(b1 %s, b2 %s)".formatted(testCaseMoney.targetRemoteType(), testCaseMoney.targetRemoteType()))) {
+            assertThatThrownBy(() -> getQueryRunner().execute(session, "SELECT a.a2, b.b2 FROM %s a JOIN %s b ON CAST(a.a1 AS %s) = b.b1".formatted(tableA.getName(), tableB.getName(), testCaseMoney.castType())));
+            // the best we can do - cast right side to a trino type, to which is resolved left side. After pushdown, trino type is mapped to native type
+            assertThat(query(session, "SELECT a.a2, b.b2 FROM %s a JOIN %s b ON CAST(a.a1 AS %s) = CAST(b.b1 AS %s)".formatted(tableA.getName(), tableB.getName(), testCaseMoney.castType(), testCaseMoney.castType())))
+                    .isFullyPushedDown();
+        }
```
Comment (ssheikin) [inline plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:1079]: @Praveen2112

File: plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java [hunk 3]
```diff
@@ -1013,6 +1017,68 @@ public void testInPredicatePushdown()
...
+        CastTestCase testCaseMoney = new CastTestCase("money", "varchar(100)", "money");
+        try (TestTable tableA = new TestTable(onRemoteDatabase(), "test_join_cast_a_", "(a1 %s, a2 %s)".formatted(testCaseMoney.fromRemoteType(), testCaseMoney.fromRemoteType()));
+                TestTable tableB = new TestTable(onRemoteDatabase(), "test_join_cast_b_", "(b1 %s, b2 %s)".formatted(testCaseMoney.targetRemoteType(), testCaseMoney.targetRemoteType()))) {
+            assertThatThrownBy(() -> getQueryRunner().execute(session, "SELECT a.a2, b.b2 FROM %s a JOIN %s b ON CAST(a.a1 AS %s) = b.b1".formatted(tableA.getName(), tableB.getName(), testCaseMoney.castType())));
+            // the best we can do - cast right side to a trino type, to which is resolved left side. After pushdown, trino type is mapped to native type
+            assertThat(query(session, "SELECT a.a2, b.b2 FROM %s a JOIN %s b ON CAST(a.a1 AS %s) = CAST(b.b1 AS %s)".formatted(tableA.getName(), tableB.getName(), testCaseMoney.castType(), testCaseMoney.castType())))
```
Comment (ssheikin) [inline plugin/trino-postgresql/src/test/java/io/trino/plugin/postgresql/TestPostgreSqlConnectorTest.java:1077]: This is a breaking change, so has to be covered by config/session property toggle.
@Praveen2112

**Comment by mosabua** [comment]
I am adding the stale-ignore label under the assumption that @ssheikin and others are still collaborating on this and will bring it to completion.

**Comment by ssheikin** [comment]
To revisit after https://github.com/trinodb/trino/pull/21607 is merged

**Pull request closed without merging**